torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,707 @@
+
+    using DescriptionKey = std::pair</*f_code=*/PyCodeObject*, /*f_lasti=*/int>;
+    ska::flat_hash_map<DescriptionKey, CodeDescription, hash_pair> code_descriptions_;
+};
+
+PythonTracer& PythonTracer::singleton() {
+    static PythonTracer singleton_;
+    return singleton_;
+}
+
+PythonTracer::PythonTracer() : active_(false) {
+    path_prefixes_ = py::module::import("torch.profiler.python_tracer")
+        .attr("_prefix_regex")().cast<std::string>();
+
+    module_call_code_ = py::module::import("torch.nn")
+        .attr("Module")
+        .attr("__call__")
+        .attr("__code__")
+        .ptr();
+}
+
+void PythonTracer::start(size_t max_threads) {
+    TORCH_CHECK(!active_, "PythonTracer is already active")
+    TORCH_CHECK(!trace_contexts_.size(), "PythonTracer should not have active contexts");
+    TORCH_CHECK(max_threads > 0, "max_threads must be positive, got ", max_threads);
+    TORCH_CHECK(
+        max_threads <= max_py_threads,
+        "max_threads must be less than or equal to ", max_py_threads);
+
+    pybind11::gil_scoped_acquire gil;
+    auto t0 = now();
+
+    // Loop over all interpreters and all threads within each interpreter.
+    // We will need to register a trace function with each thread. We set the
+    // current thread to position zero to ensure that it is traced, and so we
+    // can restore the thread state after registration.
+    std::vector<PyThreadState*> thread_states { PyThreadState_Get() };
+    if (max_threads > 1) {
+        PyInterpreterState* interpreter_state = PyInterpreterState_Head();
+        while (interpreter_state != nullptr) {
+            PyThreadState* thread_state = PyInterpreterState_ThreadHead(interpreter_state);
+            while (thread_state != nullptr) {
+                if (thread_state != thread_states[0]) {
+                    thread_states.push_back(thread_state);
+                }
+                thread_state = PyThreadState_Next(thread_state);
+            }
+            interpreter_state = PyInterpreterState_Next(interpreter_state);
+        }
+
+        if (thread_states.size() > max_threads) {
+            std::cout << "Warning: can only trace " << max_threads << " threads. "
+                    << thread_states.size() << " are currently active." << std::endl;
+            thread_states.resize(max_threads);
+        }
+    }
+
+    // Register the tracer in each thread.
+    for (size_t i = 0; i < thread_states.size(); i++) {
+        PyThreadState* thread_state = thread_states[i];
+        PyThreadState_Swap(thread_state);
+
+        auto ctx = (TraceContext*) TraceContextType.tp_alloc(&TraceContextType, 0);
+        ctx->thread_id_ = (uint8_t)i;
+        ctx->thread_state_ = thread_state;
+        ctx->initial_us_ = t0;
+        trace_contexts_.push_back(ctx);
+
+        // When we begin profiling there are already frames on the Python
+        // interpreter stack. To ensure a complete trace, we must push calls
```

Comment:
this is fine but could this cause incomplete call hierarchy ==> result in weird looking module graph?

Reply:
This is there to avoid incomplete hierarchy. (Along with this section: https://github.com/pytorch/pytorch/pull/67407/files/#diff-c45ca18f8ea99387fd472fe133defd9d3849246267f212e6d14345143e5a83fdR604-R615)

I don't expect it to interfere much with Module hierarchy since  you tend to call an entire module under the profiler, however it makes a big difference in capturing the entire Python context. (And is needed to produce a complete stack.) One thing I can do if we run into issues is add a bit or expand the enum to differentiate between an observed call/return event and a faux one added to make sure we capture everything.

Reply:
> Enabling tracing during the warmup phase will take care of this I suppose.

I don't think so. This is more about:
```
def main():
  with torch.profiler.profile(with_stack=True):
    # We already entered `main`, but we still want to record it.
    ...
```

One place where a warmup will be very useful is in the various string caches.

Reply:
Yeah ok good point if you want to capture the complete stack.